contracts/WitnetProxy.sol

```diff
@@ -0,0 +1,127 @@
+      let size := returndatasize()
+      returndatacopy(ptr, 0, size)
+      switch result
+        case 0  { 
+          // pass back revert message:
+          revert(ptr, size) 
+        }
+        default {
+          // pass back same data as returned by 'implementation' contract:
+          return(ptr, size) 
+        }
+    }
+  }
+
+  /// @dev Returns proxy's current implementation address.
+  function implementation() public view returns (address) {
+    return __proxySlot().implementation;
+  }
+
+  /// @dev Upgrades the `implementation` address.
+  /// @param _newImplementation New implementation address.
+  /// @param _initData Raw data with which new implementation will be initialized.
+  /// @return Returns whether new implementation would be further upgradable, or not.
+  function upgradeTo(address _newImplementation, bytes memory _initData)
+    public returns (bool)
+  {
+    // New implementation cannot be null:
+    require(_newImplementation != address(0), "WitnetProxy: null implementation");
+
+    address _oldImplementation = implementation();
+    if (_oldImplementation != address(0)) {
+      // New implementation address must differ from current one:
+      require(_newImplementation != _oldImplementation, "WitnetProxy: nothing to upgrade");
+
+      // Assert whether current implementation is intrinsically upgradable:
+      try Upgradable(_oldImplementation).isUpgradable() returns (bool _isUpgradable) {
+        require(_isUpgradable, "WitnetProxy: not upgradable");
+      } catch {
+        revert("WitnetProxy: unable to check upgradability");
+      }
+
+      // Assert whether current implementation allows `msg.sender` to upgrade the proxy:
+      (bool _wasCalled, bytes memory _result) = _oldImplementation.delegatecall(
+        abi.encodeWithSignature(
+          "isUpgradableFrom(address)",
+          msg.sender
+        )
+      );
+      require(_wasCalled, "WitnetProxy: not compliant");
+      require(abi.decode(_result, (bool)), "WitnetProxy: not authorized");
+      require(
+        Upgradable(_oldImplementation).proxiableUUID() == Upgradable(_newImplementation).proxiableUUID(),
+        "WitnetProxy: proxiableUUIDs mismatch"
+      );
+    }
+
+    // Initialize new implementation within proxy-context storage:
+    (bool _wasInitialized,) = _newImplementation.delegatecall(
+      abi.encodeWithSignature(
+        "initialize(bytes)",
+        _initData
+      )
+    );
+    require(_wasInitialized, "WitnetProxy: unable to initialize");
+
+    // If all checks and initialization pass, update implementation address:
+    __proxySlot().implementation = _newImplementation;
+    emit Upgraded(_newImplementation);
+
+    // Asserts new implementation complies w/ minimal implementation of Upgradable interface:
+    try Upgradable(_newImplementation).isUpgradable() returns (bool _isUpgradable) {
+      return _isUpgradable;
+    }
+    catch {
+      revert ("WitnetProxy: not compliant");
+    }
```

Comment:
What's the mechanism for renouncing upgradeability?

Reply:
Given certaint instance of this WRB implementation, the `isUpgradable()` method is to return always the same immutable value that was set on construction, no matter who calls the `isUpgradable()` method. 

This being said, for some reason you may want to deploy a WRB implementation that, while having the intrinsic capability of being upgradable, is not to be upgraded, not even by the owner address (or whatever higher-rank role is ever implemented). For covering this use-case, it's just a matter of specifying "false" as first constructor parameter when deploying the instance.

It is therefore the contract deployer the one that may renounce to upgradability on migration time.